### ISSUE TITLE
Extend `reportPrivateImportUsage` to report use of symbols imported from private submodules in a "py.typed" library

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#9008]
### Discussed in https://github.com/microsoft/pyright/discussions/9004

<div type='discussions-op-text'>

<sup>Originally posted by **Delgan** September 16, 2024</sup>
Hi.

I've read the docs about [Library interface](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface)  but I can't understand why isn't `pyright` reporting an error when I import private modules.

Consider first installing a third-party library like [`pydantic`](https://github.com/pydantic/pydantic/tree/v2.9.1/pydantic):

```bash
pip install pydantic==2.9.1
```

Then asking `pyright` to analyze the following file in "strict" mode:

```python
# pyright: strict
import pydantic._migration

print(pydantic._migration.MOVED_IN_V2)
```

Because I'm importing and using `_migration` which is supposed to be a private module, I would expect an error such as `reportPrivateUsage` or `reportPrivateImportUsage`. However, the output is (for version 1.1.380):

```bash
0 errors, 0 warnings, 0 informations 
```

Is that intentional design? I understand that for unit tests within the library, it's preferable for library authors to be able to import a private module (and test some internal functions). But shouldn't this be forbidden for a library user?</div>

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.398.

**Comment by alexmojaki** [comment]
I now have 67 errors. Is there a way to whitelist certain modules?

**Comment by erictraut** [comment]
Unfortunately, this change is causing pain for users of the popular `pandas` library because it exports most of its types from a submodule named `_typing`. This module was originally intended to be experimental, and the maintainers knew they'd eventually need to address it. They're discussing how to address it, but it will take a while. I've decided to revert this change for now to give them a chance to implement their fix and publish an updated `pandas`. This is anticipated to take about a year. I will revisit this issue at that time.

**Comment by alexmojaki** [comment]
What about configuration to whitelist certain modules?

**Comment by Andrej730** [comment]
Ahh, sorry, got confused a bit and posted to pandas repository issues instead of pyright. I'll repost the message here - maybe anyone has ideas about this?

> Is there any common practice for naming additional typing-only modules included to the stub? 
> 
> Met the similar `reportPrivateImportUsage` issue recently with `_typing` module from fake-bpy-module (https://github.com/nutti/fake-bpy-module/issues/362).
> 
> To give some background - it's stubs for [Blender Python API](https://www.blender.org/). Blender's Python API for the most part lives in C++ and there are lots of enums that API is using but they're not really exposed to Python API as such. 
> So to make stubs more readable fake-bpy-module added `_typing` module that groups all those enums using, gave them reasonable names and referred to them using `_typing` names across the API. The name `_typing` was chosen instead of `typing` to avoid giving impression to users that this module is part of the real `bpy` module.
> And this module is not for private use only, those enum literals are also can be used to statically ensure values consistency with API (in some cases they're even necessary to avoid typing errors, due to existing limitations like #8647) and without this module you would need to retype them manually and might miss some new value added in the next API update.

(posted on https://github.com/pandas-dev/pandas/issues/55231#issuecomment-2821857892)
